### DISCUSSION TITLE
Does dasel not preserve the ordering of elements within file types?

### DISCUSSION THREAD

**Discussion by ga-paul-t** [TomWright/dasel#222] [Q&A]
I have been looking around for different go libraries for setting values within different file types. I had discovered Dasel and it felt like a really great fit for a project I was working on.

However, when I use Dasel on different file types, I get unexpected behaviours (unless these are documented somewhere?):

1. YAML, value is updated but the ordering of elements is changed. Dramatically changing the original content of the file
2. TOML, value is updated, but comments are removed, a new line is added at the top of the file and the ordering of elements is changed

I haven't tried XML. But noticed there was an outstanding bug around element ordering. JSON appears to work as expected though.

**Comment by TomWright** [comment]
This is documented here: https://daseldocs.tomwright.me/notes/file-formatting-and-ordering

It's hard to deal with since maps within go are unordered by nature.

**Reply by TomWright** [reply]
I'm open to finding a fix for this, but it is almost fully down to the language parsers used.
The logic typically goes:
1. Parse file into `interface{}`, `[]interface{}`, `map[string]interface{}`
2. Perform the actual logic
3. Encode data into desired format.

These issues arise due to the fact that I'm decoding the files into generic data structures that can result in lost information specific to that format. It's a hard problem to solve but could be possible if I somehow pass additional information alongside the generic data.

**Reply by ga-paul-t** [reply]
Is it possible to use a library like https://github.com/elliotchance/orderedmap which supports the concept of an ordered map in Go? Or would that require custom parsers?

**Reply by ga-paul-t** [reply]
I should have mentioned that my use case for Dasel was to add support to a project I am building for patching semantic versions within different types of files. I just had a look at https://github.com/pelletier/go-toml and they have some utilities for converting TOML to JSON and back again. This preserves the ordering of elements (at the expense of an extra hop) but means I can use a library like sjson to set individual elements.

Feels terrible writing this, but this may be the best approach.

**Reply by TomWright** [reply]
I'll take a look to see what might be possible at the weekend.

https://github.com/pelletier/go-toml is actually the toml parser used by dasel. I may be able to adjust the data structure to include ordering information but this needs some research.

**Reply by TomWright** [reply]
FYI I'm attempting to solve this issue for all supported formats in https://github.com/TomWright/dasel/pull/289

**Reply by ga-paul-t** [reply]
> FYI I'm attempting to solve this issue for all supported formats in #289

This is awesome news!! If you need any help, happy to lend a hand

**Reply by TomWright** [reply]
I'll always accept some help
